client/src/util/stateManager/universe.js

```diff
@@ -0,0 +1,290 @@
+            "variabletype": "categorical"
+          },
+          "Cluster_2d": {
+            "displayname": "Cluster2d",
+            "include": true,
+            "type": "string",
+            "variabletype": "categorical"
+          },
+          "ERCC_reads": {
+            "displayname": "ERCC Reads",
+            "include": true,
+            "type": "int",
+            "variabletype": "continuous"
+          },
+          ...
+        }
+
+    Mapping between the two assumes:
+      - V01 only has schema for observations
+      - CellName is mapped to 'name'
+      - type conversion:   float->float32, int->int32, string->string
+
+    */
+    return {
+      annotations: {
+        obs: _.map(ota.data.schema, (val, key) => {
+          const name = key === "CellName" ? "name" : key;
+          let { type } = val;
+          if (type === "int") {
+            type = "int32";
+          }
+          if (type === "float") {
+            type = "float32";
+          }
+          return {
+            name,
+            type
+          };
+        }),
+        var: [{ name: "name", type: "string" }]
+      }
+    };
+  }
+
+  static _toObsAnnotations(ota) {
+    /*
+    v0.1 format for metadata:
+    metadata: [ { key: val, key: val, ... }, ... ]
+
+    Target format is essentially the same, except the CellName key becomes name.
+    */
+    return _.map(ota.data.metadata, (c, i) => ({
+      __obsIndex__: i,
+      name: c.CellName,
+      ...c
+    }));
+  }
+
+  static _toVarAnnotations(ota) {
+    /*
+    v0.1 initialize response contains 'genes' - names of all genes
+    in order.
+    */
+    return _.map(ota.data.genes, (g, i) => ({ __varIndex__: i, name: g }));
+  }
+
+  _toLayout(ota) {
+    /*
+    v0.1 format for the graph is:
+    [ [ 'cellname', x, y ], [ 'cellname', x, y, ], ... ]
+
+    NOTE XXX: this code does not assume any particular array ordering in the V0.1
+    response.  But for Universe initial load, the layout will be in the same
+    order as annotations, so this extra work isn't really necessary.
+    */
+
+    const { obsAnnotations } = this;
+    const obsAnnotationsByName = _.keyBy(obsAnnotations, "name");
+    const { graph } = ota.data;
+    const layout = {
+      X: new Float32Array(graph.length),
+      Y: new Float32Array(graph.length)
+    };
+
+    for (let i = 0; i < graph.length; i += 1) {
+      const [name, x, y] = graph[i];
+      const anno = obsAnnotationsByName[name];
+      const idx = anno.__obsIndex__;
+      layout.X[idx] = x;
+      layout.Y[idx] = y;
+    }
+    return layout;
+  }
+
+  _tryFinalization() {
+    if (_.every(this.init)) {
+      this._finalize();
+    }
+  }
+
+  initFromInitialize(OTAresponse) {
+    this.schema = UniverseV01._toSchema(OTAresponse);
+    this.varAnnotations = UniverseV01._toVarAnnotations(OTAresponse);
+    this.nVar = this.varAnnotations.length;
+    this.init.initialize = true;
+    this._tryFinalization();
+    return this;
```

Comment:
Prefer to avoid this pattern, open to discussion

Reply:
do you have a preferred pattern?    The order of the init calls is indeterminate, as they are the result of network calls.    So either the caller of the init functions explicitly calls something like _tryFinalization(), or the init calls do it.  I'm agnostic, so let me know what you have in mind.